modules/externalipam/ipam_helper.rb

```diff
@@ -0,0 +1,113 @@
+    no_subnets_in_group: 'No subnets found in External IPAM group',
+    provider: "The IPAM provider must be specified(e.g. 'externalipam_phpipam' or 'externalipam_netbox')",
+    groups_not_supported: 'Groups are not supported',
+    add_ip: 'Error adding IP to External IPAM',
+    bad_mac: 'Mac address is invalid',
+    bad_ip: 'IP address is invalid',
+    bad_cidr: 'The network cidr is invalid',
+  }.freeze
+
+  def provider
+    @provider ||=
+      begin
+        unless client.authenticated?
+          halt 500, { error: 'Invalid credentials for External IPAM' }.to_json
+        end
+        client
+      end
+  end
+
+  # Called when next available IP from External IPAM has been cached by another user/host, but
+  # not actually persisted in External IPAM yet. This method will increment the IP, up to
+  # MAX_IP_RETRIES times, and check if it is available in External IPAM each iteration. It
+  # will return the original IP(the 'ip' param) if no new IP's are found after MAX_IP_RETRIES
+  # iterations.
+  def find_new_ip(ip_cache, subnet_id, ip, mac, cidr, group_name)
+    found_ip = nil
+    temp_ip = ip
+    retry_count = 0
+
+    loop do
+      new_ip = increment_ip(temp_ip)
+      ipam_ip = ip_exists?(new_ip, subnet_id, group_name)
+
+      # If new IP doesn't exist in IPAM and not in the cache
+      if !ipam_ip && !ip_cache.ip_exists(new_ip, cidr, group_name)
+        found_ip = new_ip.to_s
+        ip_cache.add(found_ip, mac, cidr, group_name)
+        break
+      end
+
+      temp_ip = new_ip
+      retry_count += 1
+      break if retry_count >= MAX_IP_RETRIES
+    end
+
+    return ip if found_ip.nil?
+
+    found_ip
+  end
+
+  # Checks the cache for existing ip, and returns it if it exists. If not exists, it will
+  # find a new ip (using find_new_ip), and it is added to the cache.
+  def cache_next_ip(ip_cache, ip, mac, cidr, subnet_id, group_name)
+    group = group_name.nil? ? '' : group_name
+    ip_cache.set_group(group, {}) if ip_cache.get_group(group).nil?
+    subnet_hash = ip_cache.get_cidr(group, cidr)
+    next_ip = nil
+
+    if mac && subnet_hash&.key?(mac.to_sym)
+      next_ip = ip_cache.get_ip(group, cidr, mac)
+    else
+      new_ip = ip
+      ip_not_in_cache = subnet_hash.nil? ? true : !subnet_hash.to_s.include?(new_ip.to_s)
+
+      if ip_not_in_cache
+        next_ip = new_ip.to_s
+        ip_cache.add(new_ip, mac, cidr, group)
+      else
+        next_ip = find_new_ip(ip_cache, subnet_id, new_ip, mac, cidr, group)
+      end
+
+      unless usable_ip(next_ip, cidr)
+        return { error: "No free addresses found in subnet #{cidr}. Some available ip's may be cached. Try again in #{@ip_cache.cleanup_interval} seconds after cache is cleared." }
+      end
+    end
+
+    next_ip
+  end
+
+  def increment_ip(ip)
+    IPAddr.new(ip.to_s).succ.to_s
+  end
+
+  def usable_ip(ip, cidr)
+    network = IPAddr.new(cidr)
+    network.include?(IPAddr.new(ip)) && network.to_range.last != ip
```

Comment:
I suppose `.to_range.last` is to check the broadcast address. Technically that's IPv4 only and not relevant for IPv6. For IPv4 you would also need to check the network address.